libs/langchain-community/src/document_loaders/web/google_cloud_storage.ts

```diff
@@ -0,0 +1,99 @@
+import { Storage, StorageOptions } from '@google-cloud/storage'
+import * as os from "node:os";
+import * as path from "node:path";
+import * as fsDefault from "node:fs";
+import { BaseDocumentLoader } from "@langchain/core/document_loaders/base";
+import { Document } from "@langchain/core/documents";
+import { UnstructuredLoaderOptions, UnstructuredLoader } from "../fs/unstructured.js";
+
+
+/**
+ * Represents the parameters for the GoogleCloudStorageLoader class. It includes
+ * properties such as the GCS bucket, file destination, the options for the UnstructuredLoader and
+ * the options for Google Cloud Storage
+ */
+export type GcsLoaderConfig = {
+  fs?: typeof fsDefault;
+  bucket: string;
+  file: string;
+  unstructuredLoaderOptions: UnstructuredLoaderOptions;
+  storageOptions: StorageOptions
+}
+
+/**
+ * A class that extends the BaseDocumentLoader class. It represents a
+ * document loader for loading files from a google cloud storage bucket.
+ * @example
+ * ```typescript
+ * const loader = new GoogleCloudStorageLoader({
+ *   bucket: "<my-bucket-name>",
+ *   file: "<file-path>",
+ *   storageOptions: {
+ *     keyFilename: "<key-file-name-path>"
+ *   } 
+ *   unstructuredConfig: {
+ *     apiUrl: "<unstructured-API-URL>",
+ *     apiKey: "<unstructured-API-key>"
+ *   }
+ * });
+ * const docs = await loader.load();
+ * ```
+ */
+export class GoogleCloudStorageLoader extends BaseDocumentLoader {
+  private bucket: string;
+
+  private file: string;
+
+  private storageOptions: StorageOptions;
+
+  private _fs: typeof fsDefault;
+
+  private unstructuredLoaderOptions: UnstructuredLoaderOptions;
+
+  constructor({
+    fs = fsDefault,
+    file,
+    bucket,
+    unstructuredLoaderOptions,
+    storageOptions
+  }: GcsLoaderConfig) {
+    super();
+    this._fs = fs;
+    this.bucket = bucket;
+    this.file = file;
+    this.unstructuredLoaderOptions = unstructuredLoaderOptions;
+    this.storageOptions = storageOptions;
+  }
+
+  async load(): Promise<Document<Record<string, any>>[]> {
+    const tempDir = this._fs.mkdtempSync(
```

Comment:
nit: could we do this without the filesystem as a passthrough?

Reply:
I just follow the s3 loader solution, but I think it is possible if `UnstructuredLoader` support buffer.

I check the source code, it seems like `UnstructuredLoader` support either file path or buffer, so in theory, for s3, azure and gcp storage, we can just read the file as buffer and assign to `UnstructuredLoader`

I can take a look and make an enhancement when I have time @jacoblee93